### PR TITLE
Column-level compression block sizes

### LINKED ISSUE #54821: Per-column compression block size
> (you don't have to strictly follow this form)

**Use case**

When table has a big string column and we want to explicitly set a big compression block size for that column.

**Describe the solution you'd like**

```
CREATE TABLE t
(
        ...
        big_column String CODEC(ZSTD(1)) COMPRESSION BLOCK (min_block_size, max_block_size)
)
ENGINE = MergeTree ORDER BY tuple();
```

**Describe alternatives you've considered**

Tuning `min_compress_block_size` and `max_compress_block_size` but it's hard to get the optimal compression ratio for the big columns without using big compression block size. E.g.

```
CREATE TABLE t
(
        ... /*all small columns*/
        big_column String CODEC(ZSTD(1)) COMPRESSION BLOCK (min_block_size, max_block_size)
)
ENGINE = MergeTree ORDER BY tuple() SETTINGS min_compress_block_size = 67108864, max _compress_block_size = 67108864;
```
--> good compression ratio, but reading all columns costs too much memory.

```
CREATE TABLE t
(
        ... /*all small columns*/
        big_column String CODEC(ZSTD(1)) COMPRESSION BLOCK (min_block_size, max_block_size)
)
ENGINE = MergeTree ORDER BY tuple() SETTINGS min_compress_block_size = 16384, max _compress_block_size = 67108864;
```
---> compression ratio reduced by 30%

**Additional context**

[Greenplum](https://greenplum.org/10-examples-of-compression-enablement-in-greenplum/) can do it as:

```
CREATE TABLE t 
(
        c1 int ENCODING (compresstype=zstd),
        c2 char ENCODING (compresstype=zlib, blocksize=65536),
        c3 char
) WITH (orientation=column);
```

### PR DESCRIPTION
**Pull request by canhld94**
Closes #54821

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Certain settings (currently `min_compress_block_size` and `max_compress_block_size`) can now be specified at column-level where they take precedence over the corresponding table-level setting. Example: `CREATE TABLE tab (col String SETTINGS (min_compress_block_size = 81920, max_compress_block_size = 163840)) ENGINE = MergeTree ORDER BY tuple();`

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by canhld94** [comment]
@alexey-milovidov @nikitamikhaylov can I get some reviews for this feature :D. 
I think the syntax may be not yet the best one. However the feature itself is definitely useful.
We have tested on our prod table, the `xxxx_local` is the original one with table level compress block size and `t_local2` has same schema but with column level compress block size. The compression ratio is almost same, but memory usage is 100x time better when query table `xxxx_local2`.

```sql
┌─table─────────────┬─count()─┬─compressed_sz─┬─uncompressed_sz─┬──────────────ratio─┐
│ xxxx_html_local  │      14 │ 228.20 GiB    │ 3.43 TiB        │ 15.385512604598656 │
│ xxxx_html_local2 │      12 │ 226.07 GiB    │ 3.42 TiB        │ 15.504667251480628 │
└───────────────────┴─────────┴───────────────┴─────────────────┴────────────────────┘

-- SELECT * on origin table
SELECT * EXCEPT xxxx_html
FROM xxxx_html_local
WHERE _partition_id = '9-4-0'
SETTINGS max_threads = 16
FORMAT `Null`

Query id: 4304bfcd-a3e4-4d95-b5fa-96becee33ad0

Ok.

0 rows in set. Elapsed: 1.105 sec. Processed 5.53 million rows, 725.27 MB (5.00 million rows/s., 656.11 MB/s.)
Peak memory usage: 7.68 GiB.

-- SELECT * on new table
SELECT * EXCEPT xxxx_html
FROM xxxx_html_local2
WHERE _partition_id = '9-4-0'
SETTINGS max_threads = 16
FORMAT `Null`

Query id: 55e7290d-a6ef-4a96-badd-7569f30fb409

Ok.

0 rows in set. Elapsed: 0.172 sec. Processed 5.53 million rows, 719.40 MB (32.19 million rows/s., 4.19 GB/s.)
Peak memory usage: 33.01 MiB.
``` 

If we don't tune the table compress block size, the compression ratio (with default settings) is only ~5.x.

**Comment by alexey-milovidov** [comment]
@canhld94 The default compress block size is from 64 KB to 1 MB, and it is strange to see it could lead to the difference in 7 GB of memory usage. Is it possible that you have also changed the defaults? I don't think it's ever needed to increase the compress block size. It is one of the "factory" settings that are not expected to be changed.

**Comment by canhld94** [comment]
> @canhld94 The default compress block size is from 64 KB to 1 MB, and it is strange to see it could lead to the difference in 7 GB of memory usage. Is it possible that you have also changed the defaults?

@alexey-milovidov  May be the example is not clear. With default compress block, memory consumption is normal, but the compression ratio is not good. Previously, we need to increase table level `min_compress_block_size` and `max_compress_block_size` to 64MB, but it results in high memory consumption during SELECT queries. We tried to tune `min_compress_block_size` and `max_compress_block_size` and `index_granularity_bytes` accordingly, but none of them can achieve good compression ratio as fixed block size of 64MB.

> I don't think it's ever needed to increase the compress block size. It is one of the "factory" settings that are not expected to be changed.

In our use case, the table has a big string column (e.g. the whole html source of a website). If we use default compression block size, the compression ratio is from 5-6, which is too low for our demand.

**Comment by canhld94** [comment]
@alexey-milovidov  I've revised the example in my previous comment as well. Hope it is more clear to you.

**Comment by botsbreeder** [comment]
@alexey-milovidov we have tables where one column is a big string (100KB on average) and other columns don't have that much data. Setting min_compress_block_size to 64MB for string column helps to increase compression ratio almost twice. But if min_compress_block_size applied to the whole table all other columns use 64MB compress block and it slows down select queries and these queries require more memory. The solution is to apply min_compress_block_size to one column only and it works well in our fork (high compression ratio AND fast queries AND lower memory usage).

**Comment by canhld94** [comment]
Upgrade check #57893
Stateless `00002_log_and_exception_messages_formatting` look unrelated

**Comment by canhld94** [comment]
@alexey-milovidov we change the syntax to declare compress block size as parameters of `CODEC` to be more ClickHouse friendly. Document is also updated.

```sql
CREATE TABLE t
(
        ...
        big_column String CODEC(16384, 16384)(ZSTD(9, 24)),
)
ENGINE = MergeTree ORDER BY tuple();
```

This change is backward incompatible, so I also add a query setting `allow_declare_codec_with_parameters` that is default to false now, and we can enable (or remove it) after few stable release.

**Comment by UnamedRus** [comment]
Honestly, i preferred old syntax, it's more self describing, and also allow to support mode settings. (low_cardinality or potential dictionary support for ZSTD, i'm looking for you.)

```
        big_column String CODEC(ZSTD(9, 24)) SETTINGS (min_compress_block_size = 16384, max_compress_block_size = 16384),
```

**Comment by canhld94** [comment]
@UnamedRus yes, the old syntax is more declarative and more generic, but its scope is outside of the main purpose of this PR (to have explicit compress block size for each column). For now we want to push this PR to upstream first.

Re. column level settings, it's definitely a needed feature, but different people will have different preferred syntax and we may need lots of discussion. I still advocate my previous proposed syntax and will try to push it to upstream.

```sql
COLUMN TYPE ATRRIBUTES SETTINGS (<list of settings>), 
```

But it'll be in another issue and PR.

**Comment by rschu1ze** [comment]
@canhld94 @UnamedRus After reading this PR, #54821 and #36428, I think there is some value in per-column min/max block sizes when the columns have very different average byte sizes per value ("big string column" use case) and I like to help get this merged.

Settings in ClickHouse come in global form (configured via cfg file), session/query form (`SET ... = ...` or `SELECT ... SETTINGS`) or MergeTree settings (`CREATE TABLE ... SETTINGS ...`). Some specific settings, e.g. `min/max_compress_block_size`, exist at session and MergeTree level. This PR would make it possible for specific (not all) MergeTree settings to be overruled at column level. But we should be careful which settings we really expose at column-level, it is a balance between code complexity and benefit. `low_cardinality_max_dictionary_size` (#36428) sounds obscure to me.

Re syntax: We should strive for maximum consistency. `CODEC(min_compress_block_size, max_compress_block_size)` isn't consistent with the `SETTINGS` clause that is used elsewhere. I also agree with @UnamedRus that such syntax isn't easily extensible, it is also hard to decipher for casual users.

> `COLUMN TYPE ATRRIBUTES SETTINGS (<list of settings>)`
> But it'll be in another issue and PR.

I like that. Is there perhaps a PR already? I am afraid that if we implement `CODEC(min_compress_block_size, max_compress_block_size)`, it will be orthogonal to the better `SETTINGS` syntax, so we better start with `SYNTAX` right away.

**Comment by alexey-milovidov** [comment]
I like the syntax:
```
big_column String CODEC(ZSTD(9, 24)) SETTINGS (min_compress_block_size = 16384, max_compress_block_size = 16384),
```

Let's finish this PR and merge...

File: tests/queries/0_stateless/02870_per_column_settings.sql [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    long_string String SETTINGS (min_block_size = 81920, max_compress_block_size = 163840),
+)
+ENGINE = MergeTree
+ORDER BY id
+SETTINGS min_bytes_for_wide_part = 1; -- {serverError UNKNOWN_SETTING}
+
+SELECT '--- ';
+
+-- Column-level settings are only supported for MergeTree* tables
+CREATE TABLE tab
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02870_per_column_settings.sql:73]: This CREATE goes through successfully but an error is expected.
Did not quickly find out why, perhaps you have an idea.

File: src/Parsers/ParserCreateQuery.h [hunk 2]
```diff
@@ -321,6 +329,21 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
...
+        ParserToken parser_opening_bracket(TokenType::OpeningRoundBracket);
+        if (parser_opening_bracket.check(pos, expected))
+        {
+            if (!settings_parser.parse(pos, settings, expected))
+                return false;
+            ParserToken parser_closing_bracket(TokenType::ClosingRoundBracket);
+            if (!parser_closing_bracket.ignore(pos, expected))
+                return false;
+        }
+        /// This could be settings in alter query
```
Comment (rschu1ze) [inline src/Parsers/ParserCreateQuery.h:343]: Not clear what the comment means.

File: tests/queries/0_stateless/02870_per_column_settings.sql [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    id UInt64,
+    long_string String SETTINGS (min_compress_block_size = 163840, max_compress_block_size = 163840),
+    v1 String,
+    v2 UInt64,
+    v3 Float32,
+    v4 Float64
+)
+ENGINE = ReplicatedMergeTree('/clickhouse/tables/{database}/tab/2870', 'r1')
+ORDER BY id
+SETTINGS min_bytes_for_wide_part = 1;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02870_per_column_settings.sql:18]: Is there a reason why compact parts are not tested?

File: tests/queries/0_stateless/02870_per_column_settings.sql [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+SHOW CREATE tab;
+
+ALTER TABLE tab MODIFY COLUMN long_string RESET SETTING min_compress_block_size;
+SHOW CREATE tab;
+
+ALTER TABLE tab MODIFY COLUMN long_string REMOVE SETTINGS;
+SHOW CREATE tab;
+
+ALTER TABLE tab MODIFY COLUMN long_string String SETTINGS (min_compress_block_size = 163840, max_compress_block_size = 163840);
+SHOW CREATE tab;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02870_per_column_settings.sql:35]: Up to you but perhaps the tests should also check that column-level settings really have an effect on compression (except for a new CREATE statement returned by SHOW CREATE).

EDIT: On the other hand, column sizes differ slightly between platforms, so if this makes this test not runnable on ARM then we can level as is.

File: tests/queries/0_stateless/02870_per_column_settings.sql (hunk 3 above)
Comment (canhld94) [reply tests/queries/0_stateless/02870_per_column_settings.sql:18]: Column settings are applied for specific column, but for compact parts all columns are in a single file, so I think it column level settings doesn't make sense for compact parts.

File: tests/queries/0_stateless/02870_per_column_settings.sql (hunk 1 above)
Comment (canhld94) [reply tests/queries/0_stateless/02870_per_column_settings.sql:73]: The initial implementation only allows column settings syntax for merge tree tables. However after using internally, we think that it's better to allow the syntax for any table engines (though the settings may only take effects on merge tree). The reason is, for example we have a local table with columns settings and we want to create distributed table with same schema, then we don't need to repeat the create sql but only need `CREATE TABLE distributed AS local ENGINE = ...`.

Comment (canhld94) [reply tests/queries/0_stateless/02870_per_column_settings.sql:73]: I forgot to change the test here. Will fix it.

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**